src/AcceptanceTests/When_session_is_accessed_in_hosted_service_start.cs

```diff
@@ -0,0 +1,77 @@
+﻿namespace AcceptanceTests
+{
+    using Microsoft.Extensions.Hosting;
+    using NUnit.Framework;
+    using NServiceBus;
+    using Microsoft.Extensions.DependencyInjection;
+    using System.Threading.Tasks;
+    using System.Threading;
+    using System;
+
+    [TestFixture]
+    public class When_session_is_accessed_in_hosted_service_start
+    {
+        [Test]
+        public async Task Should_be_available_when_configured_after_NServiceBus()
+        {
+            var host = Host.CreateDefaultBuilder()
+                .UseNServiceBus(hostBuilderContext =>
+                {
+                    var endpointConfiguration = new EndpointConfiguration("NSBRepro");
+                    endpointConfiguration.SendOnly();
+                    endpointConfiguration.UseTransport<LearningTransport>();
+                    return endpointConfiguration;
+                })
+                .ConfigureServices((ctx, serviceProvider) => serviceProvider.AddHostedService<HostedServiceThatAccessSessionInStart>())
+                .Build();
+
+            await host.StartAsync();
+            await host.StopAsync();
+        }
+
+        [Test]
+        public void Should_throw_when_configured_before_NServiceBus()
+        {
+            var ex = Assert.ThrowsAsync<InvalidOperationException>(async () =>
+            {
+                var host = Host.CreateDefaultBuilder()
+                .ConfigureServices((ctx, serviceProvider) => serviceProvider.AddHostedService<HostedServiceThatAccessSessionInStart>())
+                .UseNServiceBus(hostBuilderContext =>
+                  {
+                      var endpointConfiguration = new EndpointConfiguration("NSBRepro");
+                      endpointConfiguration.SendOnly();
+                      endpointConfiguration.UseTransport<LearningTransport>();
+                      return endpointConfiguration;
+                  })
+                .Build();
+
+                await host.StartAsync();
+            });
+
+            StringAssert.Contains("The message session can't be used before the NServiceBus hosted service is started", ex.Message);
+        }
+
+        class HostedServiceThatAccessSessionInStart : IHostedService
+        {
+            public HostedServiceThatAccessSessionInStart(IServiceProvider serviceProvider)
```

Comment:
On the fence about these service provider tests. I wonder what value they bring. I see later on you different between accessing via provider or injecting into the ctor directly. But I feel the provider is just noise. What we care about is that the session can be injected an

- if accessed in the ctor of a hosted service, boom
- if accessed in hosted service which is before NServiceBus, boom
- - if accessed in hosted service which is after NServiceBus, yeah!

right?

Reply:
Yea, I had those initially until I figured that just injecting it actually accesses it, will remove those

Reply:
I've consolidated the tests now and skipped accessing them via the service provider